### DISCUSSION TITLE
Overrides not supported for deep transitive dependencies

### DISCUSSION THREAD

**Discussion by bisubus** [pnpm/pnpm#15774] [Ideas]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### Yarn resolutions

Are not supported for transitive dependencies. `typescript` entry is supported but not `@microsoft/api-extractor/typescript` or `vite-plugin-dts/@microsoft/api-extractor/typescript"`:

*package.json*
```json
{
  "devDependencies": {
    "typescript": "5.5.4",
    "vite-plugin-dts": "4.0.3"
  },
  "resolutions": {
    "vite-plugin-dts/@microsoft/api-extractor/typescript": "*"
  }
}
```

### PNPM overrides

Are supported for transitive dependencies but only 1 level deep. `@microsoft/api-extractor>typescript` is supported but not `vite-plugin-dts>@microsoft/api-extractor>typescript`:

*package.json*
```json
{
  "devDependencies": {
    "typescript": "5.5.4",
    "vite-plugin-dts": "4.0.3"
  },
  "pnpm": {
    "overrides": {
      "vite-plugin-dts>@microsoft/api-extractor>typescript": "$typescript"
    }
  }
}
```

### NPM overrides

Are not supported and ignored:

*package.json*
```json
{
  "devDependencies": {
    "typescript": "5.5.4",
    "vite-plugin-dts": "4.0.3"
  },
  "overrides": {
    "vite-plugin-dts": {
      "@microsoft/api-extractor": {
        "typescript": "$typescript"
      }
    }
  }
}
```

### Describe the Bug

`"vite-plugin-dts/@microsoft/api-extractor/typescript":` entry in Yarn `resolutions` results in an error:

>  ERR_PNPM_INVALID_SELECTOR  Cannot parse the "vite-plugin-dts/@microsoft/api-extractor/typescript" selector

`"vite-plugin-dts>@microsoft/api-extractor>typescript"` entry in PNPM `overrides` results in an error, the second `>` isn't parsed: 

> ERR_PNPM_INVALID_SELECTOR  Cannot parse the "@microsoft/api-extractor>typescript" selector

NPM `overrides` are ignored.

### Expected Behavior

Either of the suggested methods to override transitive `typescript` version is supposed to work.

Currently without overrides `pnpm list typescript --depth 10` results in the following dependency tree that contains `typescript@5.4.2`:

```
devDependencies:
typescript 5.5.4
vite-plugin-dts 4.0.3
├─┬ @microsoft/api-extractor 7.47.4
│ └── typescript 5.4.2 ❌
├─┬ @vue/language-core 2.0.29
│ └── typescript 5.5.4 peer
├── typescript 5.5.4 peer
└─┬ vue-tsc 2.0.29
  ├─┬ @vue/language-core 2.0.29
  │ └── typescript 5.5.4 peer
  └── typescript 5.5.4 peer
```

With the provided overrides it's supposed to change to `typescript@5.5.4`:

```
devDependencies:
typescript 5.5.4
vite-plugin-dts 4.0.3
├─┬ @microsoft/api-extractor 7.47.4
│ └── typescript 5.5.4 ✅
├─┬ @vue/language-core 2.0.29
│ └── typescript 5.5.4 peer
├── typescript 5.5.4 peer
└─┬ vue-tsc 2.0.29
  ├─┬ @vue/language-core 2.0.29
  │ └── typescript 5.5.4 peer
  └── typescript 5.5.4 peer
```

### Which Node.js version are you using?

20.12.2

### Which operating systems have you used?

- [ ] macOS
- [X] Windows
- [ ] Linux

**Comment by Smalyik** [comment]
Have similar issue with workspaces

_root/package.json_

```
{
  "overrides": {
      "package1@1.0.0>package2@1.0.0>mui": "2"
    }
}
```

_package2/package.json_

```
{
  "version": "1.0.0",
  "dependencies": {
    "mui": "1.0.0"
  }
}
```

_package1/package.json_

```
{
   "version": "1.0.0",
  "dependencies": {
    "mui": "2.0.0",
    "package2" : "workspaces:*"
  }
}
```

_package1/node_modules/package2/mui/package.json_

```
{
  version: "1.0.0" // overrides doesnt work
}
```

Fix with `shared-workspace-lockfile=false` and local override in package1 makes packages installation so long :(

**Comment by cheetosysst** [comment]
I'm working on a fix for this.

---
Written by an agent (Codex, GPT-6).

**Comment by cheetosysst** [comment]
I'm no longer working on this issue. I reproduced the invalid selector error, but supporting a path such as A>B>C also requires deciding how path-specific overrides behave when the same package is reached through multiple branches. That needs a resolver design decision beyond a focused bug fix, so the issue is available for someone else to pick up.

---
Written by an agent (Codex, GPT-6).
